rs/cli/src/auth.rs

```diff
@@ -125,60 +126,26 @@ impl Auth {
         }
     }
 
-    fn detect_hsm_auth() -> anyhow::Result<Option<Self>> {
-        info!("Detecting HSM devices");
-        let ctx = Pkcs11::new(Self::pkcs11_lib_path()?)?;
-        ctx.initialize(CInitializeArgs::OsThreads)?;
-        for slot in ctx.get_slots_with_token()? {
-            let info = ctx.get_slot_info(slot)?;
-            if info.slot_description().starts_with("Nitrokey Nitrokey HSM") {
-                let key_id = format!("hsm-{}-{}", info.slot_description(), info.manufacturer_id());
-                let pin_entry = Entry::new("dre-tool-hsm-pin", &key_id)?;
-                let pin = match pin_entry.get_password() {
-                    // TODO: Remove the old keyring entry search ("release-cli") after August 1st, 2024
-                    Err(Error::NoEntry) => match Entry::new("release-cli", &key_id) {
-                        Err(Error::NoEntry) => Password::new().with_prompt("Please enter the HSM PIN: ").interact()?,
-                        Ok(pin_entry) => pin_entry.get_password()?,
-                        Err(e) => return Err(anyhow::anyhow!("Failed to get pin from keyring: {}", e)),
-                    },
-                    Ok(pin) => pin,
-                    Err(e) => return Err(anyhow::anyhow!("Failed to get pin from keyring: {}", e)),
-                };
-
-                let mut flags = SessionFlags::new();
-                flags.set_serial_session(true);
-                let session = ctx.open_session_no_callback(slot, flags).unwrap();
-                session.login(UserType::User, Some(&pin))?;
-                info!("HSM login successful!");
-                pin_entry.set_password(&pin)?;
-                return Ok(Some(Auth::Hsm {
-                    pin,
-                    slot: slot.id(),
-                    key_id: "01".to_string(),
-                }));
-            }
+    pub fn create_canister_client(&self, nns_urls: Vec<url::Url>, lock: Option<Mutex<()>>) -> anyhow::Result<IcAgentCanisterClient> {
+        // FIXME: why do we even take multiple URLs if only the first one is ever used?
```

Comment:
Because how you should implement a real `Agent` is to have a pool of agents, and have fallbacks if the node you are using is unhealthy. Generally if its https://ic0.app, boundary nodes do the load balancing for you, but if its a specific node ipv6 then you are on your own.

Reply:
We should try all URLs when there's a failure in creating the canister client with the first URL.

Or actually the canister client itself should do that, like etcd or Zookeeper clients do.

Reply:
Both are correct, I would rather if agent has that logic it self, but I am not 100% sure if it does